Added support for managing custom encrypted files from cli.

morgoth opened the pull request:
One can now generate/edit file by
~`bin/rails credentials:edit config/staging-credentials.yml.enc --key config/staging.key`~
`bin/rails encrypted:edit config/staging-credentials.yml.enc --key config/staging.key`

and display it by

~`bin/rails credentials:show config/staging-credentials.yml.enc --key config/staging.key`~
`bin/rails encrypted:show config/staging-credentials.yml.enc --key config/staging.key`

When `--key` option is omitted then default `config/master.key` is used.

This is a follow up to https://github.com/rails/rails/pull/29909

/cc @dhh and @kaspth as you introduced encrypted files

rafaelfranca commented:
r? @kaspth

morgoth commented:
@kaspth Did you maybe have some time to look into it? Hopefully, we can make it to the Rails 5.2

kaspth commented:
Why do you want to use a separate credentials file for your staging environment? :)

I'm not liking a lot of the underlying code changes as they stand currently. Throwing all those configs around and manually instantiating the EncryptedConfiguration in multiple places seems off.

kaspth commented:
@dhh proposed an access API at one point that looks like:

```ruby
Rails.application.encrypted.config(:special_tokens) # Uses master key to decrypt `config/special_token.yml.enc`.
Rails.application.encrypted_with(env_key: "SPECIAL_KEY").config(:special_tokens) # With another key.
Rails.application.encrypted.file("path") # Later. Provides a wrapper for EncryptedFile. Maybe they're stored in `config/files/*.json.enc`
```

`encrypted_with` should probably just be an alias for `encrypted`, so people can pick which sounds better in a given situation.

Then `Rails.application.credentials` would just be a wrapper for `Rails.application.encrypted.config(:credentials)`.

------------------------------

So with that in mind, perhaps we shouldn't cram this in under credentials? E.g. this could work:

```bash
bin/rails encrypted:edit --config special_tokens
bin/rails encrypted:edit --file gcs_keyfile
```

dhh commented:
Don't think we even need #encrypted_with as a separate method. We could just do:

```ruby
Rails.application.encrypted(env_key: "SPECIAL_KEY").config(:special_tokens)
```

That's clear enough 👍

morgoth commented:
@kaspth For staging, I would like to have the same flow for managing credentials as for production, but obviously don't want to have all the secret values mixed into one file for both envs.

I like the idea of having separate cli `bin/rails encrypted`, so if you agree I can change this PR to introduce it. Does the following look correct?:
* `bin/rails encrypted:edit --config special_tokens` edits file `config/special_tokens.yml.enc`
* `bin/rails encrypted:edit --config special_tokens --key config/special.key` uses the given key defaulting to `config/master.key`
* `Rails.application.encrypted(env_key: "SPECIAL_KEY").config(:special_tokens)` decrypts `config/special_tokens.yml.enc` using env `SPECIAL_KEY`

kaspth commented:
Had a chat with @dhh 

This is the API we'd like:

```ruby
Rails.application.encrypted("config/special_tokens.yml.enc", env_key: "SOMETHING")
```

And for the CLI:

```
bin/rails encrypted:edit config/special_tokens.yml.enc --key config/something.key
```

morgoth committed:
Added CLI for encrypted files management.

One can edit/show encrypted file by:
```
bin/rails encrypted:edit config/staging_tokens.yml.enc
bin/rails encrypted:edit config/staging_tokens.yml.enc --key config/staging.key
bin/rails encrypted:show config/staging_tokens.yml.enc
```

morgoth commented:
@kaspth I added CLI wrapper for `ActiveSupport::EncryptedFile` to not force `YAML` format (which is used in `ActiveSupport::EncryptedConfiguration`).
If that's ok, then having `Rails.application.encrypted` doesn't give that much as one would have to manually cast it to given format like `JSON.load(Rails.application.encrypted("config/file.json.enc")`, which is not any better than `JSON.load(ActiveSupport::EncryptedFile.new(content_path: "config/file.json.enc").read)`

Should we leave it like this or change the CLI to be a wrapper for `ActiveSupport::EncryptedConfiguration` which is strictly a `YAML` file?

kaspth commented:
`Rails.application.encrypted` gives us plenty in readability. Plus the manual encrypted file instantiation is missing `config_path`, `key_path` and/or `env_key` arguments, so it's not the same comparison.

`Rails.application.encrypted` should just be a wrapper for the encrypted configuration then calling `read` would just interpret it as an encrypted file while `config` would `read` it and parse YAML.

kaspth commented:
Glancing over the code, I think this would take too many review rounds to get it to where I'd like it to be. It also showed that there's some kinks in the CLI API that needs to be thought through, which I don't think I can do without playing around locally.

Mind if I take over? I'd base it on your work and you'd still get credit on contributors.rubyonrails.org :)

morgoth commented:
Of course. Go ahead :-)

kaspth commented:
@morgoth got it in there, thanks! ❤️

arianf commented:
> Why do you want to use a separate credentials file for your staging environment? :)

The answer to this question seems very obvious... you don't want to encrypt your production passwords with the same encryption key as your staging env. For security each environment should have its own encryption key.
